File: frontend/js/services.js
/**
 * Services Management Module
 * Handles service CRUD operations and OpenAPI spec management
 */

/**
 * Load and display services list
 */
async function loadServices() {
  showLoading("services-list");

  try {
    const response = await apiRequest("/services");
    const servicesData = response.data || [];

    if (servicesData.length === 0) {
      showEmpty(
        "services-list",
        "No services found",
        '<div class="mt-2"><button class="btn btn-primary btn-sm" onclick="showCreateServiceModal()"><i class="fas fa-plus me-1"></i>Create First Service</button></div>'
      );
      return;
    }

    displayServicesTable(servicesData);
  } catch (error) {
    showError("services-list", "Failed to load services");
  }
}

/**
 * Display services in a table format
 */
function displayServicesTable(services) {
  const container = document.getElementById("services-list");

  const table = `
        <table class="table table-hover">
            <thead>
                <tr>
                    <th>Service Name</th>
                    <th>Status</th>
                    <th>Endpoints</th>
                    <th>Test Cases</th>
                    <th>Test Data</th>
                    <th>Created</th>
                    <th>Actions</th>
                </tr>
            </thead>
            <tbody>
                ${services
                  .map(
                    (service) => `
                    <tr>
                        <td>
                            <div>
                                <strong>${escapeHtml(service.name)}</strong>
                                <br>
                                <small class="text-muted">${service.id}</small>
                            </div>
                        </td>
                        <td>
                            <span class="badge ${getStatusBadgeClass(
                              service.status
                            )}">${service.status}</span>
                        </td>
                        <td>
                            <span class="badge bg-primary">${
                              service.endpoints_count
                            }</span>
                        </td>
                        <td>
                            <span class="badge bg-success">${
                              service.test_cases_count
                            }</span>
                        </td>
                        <td>
                            <span class="badge bg-info">${
                              service.test_data_count
                            }</span>
                        </td>
                        <td>
                            <small>${formatDate(service.created_at)}</small>
                        </td>
                        <td>
                            <div class="btn-group" role="group">
                                <button class="btn btn-sm btn-outline-primary" onclick="viewServiceDetails('${
                                  service.id
                                }')" title="View Details">
                                    <i class="fas fa-eye"></i>
                                </button>
                                <button class="btn btn-sm btn-outline-secondary" onclick="viewServiceEndpoints('${
                                  service.id
                                }')" title="View Endpoints">
                                    <i class="fas fa-list"></i>
                                </button>
                                <button class="btn btn-sm btn-outline-warning" onclick="editServiceSpec('${
                                  service.id
                                }')" title="Edit Spec">
                                    <i class="fas fa-edit"></i>
                                </button>
                                <button class="btn btn-sm btn-outline-danger" onclick="deleteService('${
                                  service.id
                                }', '${escapeHtml(
                      service.name
                    )}')" title="Delete">
                                    <i class="fas fa-trash"></i>
                                </button>
                            </div>
                        </td>
                    </tr>
                `
                  )
                  .join("")}
            </tbody>
        </table>
    `;

  container.innerHTML = table;
}

/**
 * Show create service modal
 */
function showCreateServiceModal() {
  // Reset form
  document.getElementById("createServiceForm").reset();
  document.getElementById("specSourceType").value = "upload";
  handleSpecSourceChange();

  showModal("createServiceModal");
}

/**
 * Handle spec source type change
 */
function handleSpecSourceChange() {
  const sourceType = document.getElementById("specSourceType").value;

  // Hide all divs
  document.getElementById("uploadSpecDiv").style.display = "none";
  document.getElementById("existingSpecDiv").style.display = "none";
  document.getElementById("urlSpecDiv").style.display = "none";

  // Show relevant div
  switch (sourceType) {
    case "upload":
      document.getElementById("uploadSpecDiv").style.display = "block";
      break;
    case "existing":
      document.getElementById("existingSpecDiv").style.display = "block";
      break;
    case "url":
      document.getElementById("urlSpecDiv").style.display = "block";
      break;
  }
}

/**
 * Create new service
 */
async function createService() {
  try {
    const form = document.getElementById("createServiceForm");
    const formData = new FormData(form);

    const serviceName = document.getElementById("serviceName").value.trim();
    const sourceType = document.getElementById("specSourceType").value;
    const rebuildOdg = document.getElementById("rebuildOdg").checked;

    if (!serviceName) {
      showNotification("Please enter a service name", "warning");
      return;
    }

    let pathOrUrl = "";

    // Handle different source types
    if (sourceType === "upload") {
      const fileInput = document.getElementById("specFile");
      if (!fileInput.files[0]) {
        showNotification("Please select a file to upload", "warning");
        return;
      }

      // First upload the file
      pathOrUrl = await uploadSpecFile(fileInput.files[0]);
    } else if (sourceType === "existing") {
      pathOrUrl = document.getElementById("existingSpecPath").value.trim();
      if (!pathOrUrl) {
        showNotification(
          "Please enter the path to existing spec file",
          "warning"
        );
        return;
      }
    } else if (sourceType === "url") {
      pathOrUrl = document.getElementById("specUrl").value.trim();
      if (!pathOrUrl) {
        showNotification("Please enter the spec URL", "warning");
        return;
      }
    }

    // Create service
    const serviceData = {
      service_name: serviceName,
      swagger_source: {
        type: sourceType,
        path_or_url: pathOrUrl,
      },
      rebuild_odg: rebuildOdg,
    };

    const response = await apiRequest("/services", {
      method: "POST",
      body: JSON.stringify(serviceData),
    });

    showNotification("Service created successfully!", "success");
    hideModal("createServiceModal");
    loadServices(); // Refresh services list
  } catch (error) {
    showNotification("Failed to create service: " + error.message, "danger");
  }
}

/**
 * Upload spec file
 */
async function uploadSpecFile(file) {
  const formData = new FormData();
  formData.append("file", file);

  try {
    const response = await fetch(`${API_BASE_URL}/services/upload-spec`, {
      method: "POST",
      body: formData,
    });

    const data = await response.json();

    if (!response.ok) {
      throw new Error(data.detail || "Failed to upload file");
    }

    return data.data.file_path;
  } catch (error) {
    console.error("File upload failed:", error);
    throw error;
  }
}

/**
 * View service details
 */
async function viewServiceDetails(serviceId) {
  try {
    const response = await apiRequest(`/services/${serviceId}`);
    const service = response.data;

    // Create modal dynamically
    const modalHtml = `
            <div class="modal fade" id="serviceDetailsModal" tabindex="-1">
                <div class="modal-dialog modal-lg">
                    <div class="modal-content">
                        <div class="modal-header">
                            <h5 class="modal-title">Service Details: ${escapeHtml(
                              service.name
                            )}</h5>
                            <button type="button" class="btn-close" data-bs-dismiss="modal"></button>
                        </div>
                        <div class="modal-body">
                            <div class="row">
                                <div class="col-md-6">
                                    <h6>Basic Information</h6>
                                    <table class="table table-sm">
                                        <tr><td><strong>ID:</strong></td><td>${
                                          service.id
                                        }</td></tr>
                                        <tr><td><strong>Name:</strong></td><td>${escapeHtml(
                                          service.name
                                        )}</td></tr>
                                        <tr><td><strong>Status:</strong></td><td><span class="badge ${getStatusBadgeClass(
                                          service.status
                                        )}">${service.status}</span></td></tr>
                                        <tr><td><strong>Created:</strong></td><td>${formatDate(
                                          service.created_at
                                        )}</td></tr>
                                        <tr><td><strong>Updated:</strong></td><td>${formatDate(
                                          service.updated_at
                                        )}</td></tr>
                                    </table>
                                </div>
                                <div class="col-md-6">
                                    <h6>Statistics</h6>
                                    <table class="table table-sm">
                                        <tr><td><strong>Endpoints:</strong></td><td><span class="badge bg-primary">${
                                          service.endpoints_count
                                        }</span></td></tr>
                                        <tr><td><strong>Test Cases:</strong></td><td><span class="badge bg-success">${
                                          service.test_cases_count
                                        }</span></td></tr>
                                        <tr><td><strong>Test Data:</strong></td><td><span class="badge bg-info">${
                                          service.test_data_count
                                        }</span></td></tr>
                                    </table>
                                </div>
                            </div>
                            <div class="row mt-3">
                                <div class="col-12">
                                    <h6>File Paths</h6>
                                    <table class="table table-sm">
                                        <tr><td><strong>Spec File:</strong></td><td><code>${
                                          service.spec_path
                                        }</code></td></tr>
                                        <tr><td><strong>Working Directory:</strong></td><td><code>${
                                          service.working_dir
                                        }</code></td></tr>
                                        <tr><td><strong>Spec Source:</strong></td><td>${
                                          service.spec_source
                                        }</td></tr>
                                    </table>
                                </div>
                            </div>
                        </div>
                        <div class="modal-footer">
                            <button type="button" class="btn btn-primary" onclick="viewServiceEndpoints('${
                              service.id
                            }')">
                                <i class="fas fa-list me-2"></i>View Endpoints
                            </button>
                            <button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Close</button>
                        </div>
                    </div>
                </div>
            </div>
        `;

    // Remove existing modal if any
    const existingModal = document.getElementById("serviceDetailsModal");
    if (existingModal) {
      existingModal.remove();
    }

    // Add new modal to DOM
    document.body.insertAdjacentHTML("beforeend", modalHtml);
    showModal("serviceDetailsModal");
  } catch (error) {
    showNotification(
      "Failed to load service details: " + error.message,
      "danger"
    );
  }
}

/**
 * View service endpoints
 */
async function viewServiceEndpoints(serviceId) {
  try {
    const response = await apiRequest(`/services/${serviceId}/endpoints`);
    const endpoints = response.data || [];

    // Create modal dynamically
    const modalHtml = `
            <div class="modal fade" id="serviceEndpointsModal" tabindex="-1">
                <div class="modal-dialog modal-lg">
                    <div class="modal-content">
                        <div class="modal-header">
                            <h5 class="modal-title">Service Endpoints</h5>
                            <button type="button" class="btn-close" data-bs-dismiss="modal"></button>
                        </div>
                        <div class="modal-body">
                            ${
                              endpoints.length === 0
                                ? '<div class="text-center text-muted py-4"><i class="fas fa-inbox me-2"></i>No endpoints found</div>'
                                : `<div class="list-group">
                                    ${endpoints
                                      .map(
                                        (endpoint) => `
                                        <div class="list-group-item">
                                            <div class="d-flex justify-content-between align-items-center">
                                                <div>
                                                    <span class="endpoint-method ${getMethodBadgeClass(
                                                      endpoint.method
                                                    )}">${
                                          endpoint.method
                                        }</span>
                                                    <code>${
                                                      endpoint.path
                                                    }</code>
                                                </div>
                                                <div>
                                                    <small class="text-muted">${
                                                      endpoint.operation_id
                                                    }</small>
                                                </div>
                                            </div>
                                            ${
                                              endpoint.summary
                                                ? `<small class="text-muted">${escapeHtml(
                                                    endpoint.summary
                                                  )}</small>`
                                                : ""
                                            }
                                        </div>
                                    `
                                      )
                                      .join("")}
                                </div>`
                            }
                        </div>
                        <div class="modal-footer">
                            <button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Close</button>
                        </div>
                    </div>
                </div>
            </div>
        `;

    // Remove existing modal if any
    const existingModal = document.getElementById("serviceEndpointsModal");
    if (existingModal) {
      existingModal.remove();
    }

    // Add new modal to DOM
    document.body.insertAdjacentHTML("beforeend", modalHtml);
    showModal("serviceEndpointsModal");
  } catch (error) {
    showNotification("Failed to load endpoints: " + error.message, "danger");
  }
}

/**
 * Edit service specification
 */
async function editServiceSpec(serviceId) {
  try {
    const response = await apiRequest(`/services/${serviceId}`);
    const service = response.data;

    // Create modal dynamically
    const modalHtml = `
            <div class="modal fade" id="editSpecModal" tabindex="-1">
                <div class="modal-dialog modal-xl">
                    <div class="modal-content">
                        <div class="modal-header">
                            <h5 class="modal-title">Edit OpenAPI Specification: ${escapeHtml(
                              service.name
                            )}</h5>
                            <button type="button" class="btn-close" data-bs-dismiss="modal"></button>
                        </div>
                        <div class="modal-body">
                            <div class="mb-3">
                                <label class="form-label">OpenAPI Specification (JSON)</label>
                                <textarea class="form-control" id="specContent" rows="20" style="font-family: monospace;"></textarea>
                            </div>
                            <div class="mb-3">
                                <div class="form-check">
                                    <input class="form-check-input" type="checkbox" id="rebuildOdgEdit" checked>
                                    <label class="form-check-label" for="rebuildOdgEdit">
                                        Rebuild Operation Dependency Graph (ODG)
                                    </label>
                                </div>
                            </div>
                        </div>
                        <div class="modal-footer">
                            <button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Cancel</button>
                            <button type="button" class="btn btn-primary" onclick="updateServiceSpec('${serviceId}')">
                                <i class="fas fa-save me-2"></i>Update Specification
                            </button>
                        </div>
                    </div>
                </div>
            </div>
        `;

    // Remove existing modal if any
    const existingModal = document.getElementById("editSpecModal");
    if (existingModal) {
      existingModal.remove();
    }

    // Add new modal to DOM
    document.body.insertAdjacentHTML("beforeend", modalHtml);

    // Load current spec content
    try {
      const specResponse = await fetch(service.spec_path);
      const specContent = await specResponse.text();

      // Try to format JSON
      try {
        const parsed = JSON.parse(specContent);
        document.getElementById("specContent").value = JSON.stringify(
          parsed,
          null,
          2
        );
      } catch {
        document.getElementById("specContent").value = specContent;
      }
    } catch {
      document.getElementById("specContent").value =
        "// Unable to load current specification";
    }

    showModal("editSpecModal");
  } catch (error) {
    showNotification(
      "Failed to load service for editing: " + error.message,
      "danger"
    );
  }
}

/**
 * Update service specification
 */
async function updateServiceSpec(serviceId) {
  try {
    const specContent = document.getElementById("specContent").value.trim();
    const rebuildOdg = document.getElementById("rebuildOdgEdit").checked;

    if (!specContent) {
      showNotification("Please enter the specification content", "warning");
      return;
    }

    // Validate JSON
    try {
      JSON.parse(specContent);
    } catch (error) {
      showNotification(
        "Invalid JSON specification: " + error.message,
        "warning"
      );
      return;
    }

    const updateData = {
      spec_content: specContent,
      rebuild_odg: rebuildOdg,
    };

    await apiRequest(`/services/${serviceId}/spec`, {
      method: "PUT",
      body: JSON.stringify(updateData),
    });

    showNotification("Service specification updated successfully!", "success");
    hideModal("editSpecModal");
    loadServices(); // Refresh services list
  } catch (error) {
    showNotification(
      "Failed to update specification: " + error.message,
      "danger"
    );
  }
}

/**
 * Delete service
 */
async function deleteService(serviceId, serviceName) {
  if (
    !confirm(
      `Are you sure you want to delete the service "${serviceName}"?\n\nThis will permanently delete all associated test cases, test data, and results.`
    )
  ) {
    return;
  }

  try {
    await apiRequest(`/services/${serviceId}`, {
      method: "DELETE",
    });

    showNotification("Service deleted successfully!", "success");
    loadServices(); // Refresh services list
  } catch (error) {
    showNotification("Failed to delete service: " + error.message, "danger");
  }
}

// Export functions for global access
window.loadServices = loadServices;
window.showCreateServiceModal = showCreateServiceModal;
window.handleSpecSourceChange = handleSpecSourceChange;
window.createService = createService;
window.viewServiceDetails = viewServiceDetails;
window.viewServiceEndpoints = viewServiceEndpoints;
window.editServiceSpec = editServiceSpec;
window.updateServiceSpec = updateServiceSpec;
window.deleteService = deleteService;

File: server/routes/services.py
"""
Services API routes for managing OpenAPI specifications and services
"""

from fastapi import APIRouter, Request, HTTPException, UploadFile, File
from typing import List
import json
import os
import sys
from pathlib import Path

# Add parent directory to path
sys.path.append(str(Path(__file__).parent.parent.parent))

from ..models import (
    CreateServiceRequest, ServiceSummary, ServiceDetail, UpdateServiceSpecRequest,
    ApiResponse, EndpointInfo, SchemaInfo
)
from ..database import DatabaseManager
from ..services.integration import KATIntegrationService

router = APIRouter()


@router.post("/services", response_model=ApiResponse)
async def create_service(request: Request, service_data: CreateServiceRequest):
    """Create a new service with OpenAPI specification"""
    try:
        db_manager: DatabaseManager = request.app.state.db
        
        # Handle different spec source types
        spec_content = None
        if service_data.swagger_source.type == "upload":
            # Read spec from uploaded file path
            spec_path = service_data.swagger_source.path_or_url
            if not os.path.exists(spec_path):
                raise HTTPException(status_code=400, detail="Spec file not found")
            
            with open(spec_path, 'r', encoding='utf-8') as f:
                spec_content = f.read()
                
        elif service_data.swagger_source.type == "url":
            # TODO: Implement URL fetching
            raise HTTPException(status_code=501, detail="URL spec source not implemented yet")
            
        elif service_data.swagger_source.type == "existing":
            # Use existing file
            spec_path = service_data.swagger_source.path_or_url
            if not os.path.exists(spec_path):
                raise HTTPException(status_code=400, detail="Existing spec file not found")
                
            with open(spec_path, 'r', encoding='utf-8') as f:
                spec_content = f.read()
        
        # Create service in database
        service_id = db_manager.create_service(
            service_name=service_data.service_name,
            spec_content=spec_content,
            spec_source=service_data.swagger_source.path_or_url
        )
        
        # Setup KAT integration  
        integration_service = KATIntegrationService(
            service_id=service_id,
            service_name=service_data.service_name
        )
        
        # Setup KAT-compatible directory structure
        kat_paths = integration_service.setup_service(spec_content)
        
        # Update service with endpoints count
        db_manager.update_service(service_id, {
            "endpoints_count": len(integration_service.get_endpoints_from_spec())
        })
        
        # TODO: If rebuild_odg is True, trigger ODG generation
        if service_data.rebuild_odg:
            # This will be implemented when we integrate with ODG generator
            pass
        
        return ApiResponse(
            success=True,
            message=f"Service '{service_data.service_name}' created successfully",
            data={"service_id": service_id}
        )
        
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to create service: {str(e)}")


@router.get("/services", response_model=ApiResponse)
async def list_services(request: Request):
    """List all registered services"""
    try:
        db_manager: DatabaseManager = request.app.state.db
        services_data = db_manager.get_all_services()
        
        services = [
            ServiceSummary(
                id=service["id"],
                name=service["name"],
                status=service["status"],
                endpoints_count=service.get("endpoints_count", 0),
                test_cases_count=service.get("test_cases_count", 0),
                test_data_count=service.get("test_data_count", 0),
                created_at=service["created_at"],
                updated_at=service["updated_at"]
            )
            for service in services_data
        ]
        
        return ApiResponse(
            success=True,
            message="Services retrieved successfully",
            data=services
        )
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to retrieve services: {str(e)}")


@router.get("/services/{service_id}", response_model=ApiResponse)
async def get_service(request: Request, service_id: str):
    """Get detailed information about a specific service"""
    try:
        db_manager: DatabaseManager = request.app.state.db
        service_data = db_manager.get_service(service_id)
        
        if not service_data:
            raise HTTPException(status_code=404, detail="Service not found")
        
        service_detail = ServiceDetail(
            id=service_data["id"],
            name=service_data["name"],
            status=service_data["status"],
            endpoints_count=service_data.get("endpoints_count", 0),
            test_cases_count=service_data.get("test_cases_count", 0),
            test_data_count=service_data.get("test_data_count", 0),
            created_at=service_data["created_at"],
            updated_at=service_data["updated_at"],
            spec_path=service_data["spec_path"],
            spec_source=service_data["spec_source"],
            working_dir=service_data["working_dir"]
        )
        
        return ApiResponse(
            success=True,
            message="Service details retrieved successfully",
            data=service_detail
        )
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to retrieve service: {str(e)}")


@router.put("/services/{service_id}/spec", response_model=ApiResponse)
async def update_service_spec(request: Request, service_id: str, spec_data: UpdateServiceSpecRequest):
    """Update OpenAPI specification for a service"""
    try:
        db_manager: DatabaseManager = request.app.state.db
        service_data = db_manager.get_service(service_id)
        
        if not service_data:
            raise HTTPException(status_code=404, detail="Service not found")
        
        # Update spec file
        spec_path = service_data["spec_path"]
        with open(spec_path, 'w', encoding='utf-8') as f:
            if isinstance(spec_data.spec_content, str):
                # Validate JSON
                json.loads(spec_data.spec_content)
                f.write(spec_data.spec_content)
            else:
                json.dump(spec_data.spec_content, f, indent=2)
        
        # Update database
        db_manager.update_service(service_id, {
            "updated_at": db_manager.load_data()["services"][service_id]["updated_at"]
        })
        
        # TODO: If rebuild_odg is True, trigger ODG regeneration
        if spec_data.rebuild_odg:
            pass
        
        return ApiResponse(
            success=True,
            message="Service specification updated successfully"
        )
        
    except json.JSONDecodeError:
        raise HTTPException(status_code=400, detail="Invalid JSON specification")
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to update specification: {str(e)}")


@router.delete("/services/{service_id}", response_model=ApiResponse)
async def delete_service(request: Request, service_id: str):
    """Delete a service and all its associated data"""
    try:
        db_manager: DatabaseManager = request.app.state.db
        
        if not db_manager.get_service(service_id):
            raise HTTPException(status_code=404, detail="Service not found")
        
        success = db_manager.delete_service(service_id)
        
        if success:
            return ApiResponse(
                success=True,
                message="Service deleted successfully"
            )
        else:
            raise HTTPException(status_code=500, detail="Failed to delete service")
            
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to delete service: {str(e)}")

@router.delete("/services/by-name/{service_name}", response_model=ApiResponse)
async def delete_service(request: Request, service_name: str):
    """Delete a service and all its associated data"""
    try:
        db_manager: DatabaseManager = request.app.state.db
        
        # Find service by name
        services = db_manager.get_all_services()
        service = next((s for s in services if s["name"] == service_name), None)
        
        if not service:
            raise HTTPException(status_code=404, detail="Service not found")
        
        success = db_manager.delete_service_by_name(service_name)
        
        if success:
            return ApiResponse(
                success=True,
                message="Service deleted successfully"
            )
        else:
            raise HTTPException(status_code=500, detail="Failed to delete service")
            
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to delete service: {str(e)}")
@router.get("/services/{service_id}/endpoints", response_model=ApiResponse)
async def get_service_endpoints(request: Request, service_id: str):
    """Get list of endpoints for a service"""
    try:
        db_manager: DatabaseManager = request.app.state.db
        service_data = db_manager.get_service(service_id)
        
        if not service_data:
            raise HTTPException(status_code=404, detail="Service not found")
        
        # Use KAT integration to extract endpoints
        integration_service = KATIntegrationService(
            service_id=service_id,
            service_name=service_data["name"]
        )
        
        endpoints_data = integration_service.get_endpoints_from_spec()
        
        endpoints = [
            EndpointInfo(
                method=endpoint["method"],
                path=endpoint["path"],
                operation_id=endpoint["operation_id"],
                summary=f"{endpoint['method']} {endpoint['path']}"
            )
            for endpoint in endpoints_data
        ]
        
        return ApiResponse(
            success=True,
            message="Endpoints retrieved successfully",
            data=endpoints
        )
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to retrieve endpoints: {str(e)}")


@router.get("/services/{service_id}/schemas", response_model=ApiResponse)
async def get_service_schemas(request: Request, service_id: str):
    """Get schemas from OpenAPI specification"""
    try:
        db_manager: DatabaseManager = request.app.state.db
        service_data = db_manager.get_service(service_id)
        
        if not service_data:
            raise HTTPException(status_code=404, detail="Service not found")
        
        # Use KAT integration to extract schemas
        integration_service = KATIntegrationService(
            service_id=service_id,
            service_name=service_data["name"]
        )
        
        schemas_data = integration_service.get_schemas_from_spec()
        
        schemas = [
            SchemaInfo(
                name=name,
                type=schema.get("type", "object"),
                properties=schema.get("properties", {})
            )
            for name, schema in schemas_data.items()
        ]
        
        return ApiResponse(
            success=True,
            message="Schemas retrieved successfully",
            data=schemas
        )
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to retrieve schemas: {str(e)}")


@router.post("/services/upload-spec", response_model=ApiResponse)
async def upload_spec_file(file: UploadFile = File(...)):
    """Upload OpenAPI specification file"""
    try:
        if not file.filename.endswith(('.json', '.yaml', '.yml')):
            raise HTTPException(status_code=400, detail="Only JSON and YAML files are supported")
        
        # Save uploaded file temporarily
        project_root = Path(__file__).parent.parent.parent
        uploads_dir = project_root / "server" / "uploads"
        uploads_dir.mkdir(exist_ok=True)
        
        file_path = uploads_dir / file.filename
        
        with open(file_path, 'wb') as f:
            content = await file.read()
            f.write(content)
        
        # Validate the file content
        if file.filename.endswith('.json'):
            with open(file_path, 'r', encoding='utf-8') as f:
                json.load(f)  # Validate JSON
        
        return ApiResponse(
            success=True,
            message="Specification file uploaded successfully",
            data={"file_path": str(file_path)}
        )
        
    except json.JSONDecodeError:
        raise HTTPException(status_code=400, detail="Invalid JSON file")
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to upload file: {str(e)}")
